crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+use std::iter::FusedIterator;
+
+use ruff_db::file_system::{FileSystemPath, FileSystemPathBuf};
+use ruff_db::vfs::{system_path_to_file, VfsPath};
+
+use crate::module_name::ModuleName;
+use crate::supported_py_version::get_target_py_version;
+use crate::typeshed::{parse_typeshed_versions, TypeshedVersionsQueryResult};
+use crate::{Db, TypeshedVersions};
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct ExtraPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct ExtraPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct FirstPartyPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct FirstPartyPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct StandardLibraryPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct StandardLibraryPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct SitePackagesPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct SitePackagesPathBuf(FileSystemPathBuf);
+
+/// Enumeration of the different kinds of search paths type checkers are expected to support.
+///
+/// N.B. Although we don't implement `Ord` for this enum, they are ordered in terms of the
+/// priority that we want to give these modules when resolving them,
+/// as per [the order given in the typing spec]
+///
+/// [the order given in the typing spec]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum ModuleResolutionPath {
```

Comment:
I think I would call this enum `ModuleResolutionPathBuf` and the other `ModuleResolutionPath`

Reply:
Hmm, I made this change, but now I'm wondering if that naming scheme is somewhat confusing. `FooPath` structs generally always need to be behind `&` references when you use them, because they're unsized types. But it's impossible to have an unsized enum (unfortunately), so the enum that represents the borrowed version of the path cannot be an unsized type; instead, it holds references internally.

Reply:
The owned enum is now called `ModuleResolutionPathBuf`, as per your suggestion, but the one that has borrowed variants I've renamed back to `ModuleResolutionPathRef`